chain/consensus/tendermint/mine.go

```diff
@@ -0,0 +1,244 @@
+	}
+
+	nn, err := api.StateNetworkName(ctx)
+	if err != nil {
+		return err
+	}
+	subnetID := address.SubnetID(nn)
+
+	log.Infof("miner params: network:%s, subnet ID: %s", nn, subnetID)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		base, err := api.ChainHead(ctx)
+		if err != nil {
+			log.Errorw("getting block failed", "error", err)
+			continue
+		}
+
+		log.Infof("[%s] epoch is %d", subnetID, base.Height())
+
+		msgs, err := api.MpoolSelect(ctx, base.Key(), 1)
+		if err != nil {
+			log.Errorw("unable to select messages from mempool", "error", err)
+		}
+
+		crossMsgs, err := api.GetCrossMsgsPool(ctx, subnetID, base.Height()+1)
+		if err != nil {
+			log.Errorw("selecting cross-messages failed", "error", err)
+		}
+		log.Infof("[%s] retrieved %d - msgs, %d - crossmsgs from the pool for @%s", subnetID, len(msgs), len(crossMsgs), base.Height()+1)
+
+		for _, msg := range msgs {
+			id := msg.Cid().String()
+
+			if cache.shouldSendMessage(id) {
+				msgBytes, err := msg.Serialize()
+				if err != nil {
+					log.Error(err)
+					continue
+				}
+				tx := NewSignedMessageBytes(msgBytes)
+				_, err = tendermintClient.BroadcastTxSync(ctx, tx)
+				if err != nil {
+					log.Error("unable to send a message to Tendermint:", err)
+					continue
+				} else {
+					cache.addSentMessage(id, base.Height())
+					log.Infof("successfully sent a message with ID %s to Tendermint", id)
+				}
+			}
+		}
+
+		for _, msg := range crossMsgs {
+			id := msg.Cid().String()
+
+			if cache.shouldSendMessage(id) {
+				msgBytes, err := msg.Serialize()
+				if err != nil {
+					log.Error(err)
+					continue
+				}
+				tx := NewCrossMessageBytes(msgBytes)
+				_, err = tendermintClient.BroadcastTxSync(ctx, tx)
+				if err != nil {
+					log.Error("unable to send a cross message to Tendermint:", err)
+					continue
+				} else {
+					cache.addSentMessage(id, base.Height())
+					log.Infof("successfully sent a message with ID %s to Tendermint", id)
+				}
+			}
+		}
+
+		log.Infof("[%s] try creating Tendermint block for epoch %d", subnetID, base.Height()+1)
+
+		bh, err := api.MinerCreateBlock(ctx, &lapi.BlockTemplate{
+			Miner:            address.Undef,
+			Parents:          base.Key(),
+			BeaconValues:     nil,
+			Ticket:           nil,
+			Epoch:            base.Height() + 1,
+			Timestamp:        0,
+			WinningPoStProof: nil,
+			Messages:         nil,
+			CrossMessages:    nil,
+		})
+		if err != nil {
+			log.Errorw("creating block failed", "error", err)
+			continue
+		}
+		if bh == nil {
+			continue
+		}
+
+		log.Infof("[%s] try syncing Tendermint block for epoch %d", subnetID, base.Height()+1)
+
+		err = api.SyncSubmitBlock(ctx, &types.BlockMsg{
+			Header:        bh.Header,
+			BlsMessages:   bh.BlsMessages,
+			SecpkMessages: bh.SecpkMessages,
+			CrossMessages: bh.CrossMessages,
+		})
+		if err != nil {
+			log.Errorw("submitting block failed", "error", err)
+		}
+
+		cache.clearSentMessages(base.Height())
```

Comment:
I would prefer to address this later. Because it will complicate the implementation. The cache is just a map now and for a production-ready system it should be something else, which means that we will have to rewrite it anyway. So let's defer for the future.

Reply:
Can you add a comment (with NOTE or FIXME) so it is easy to locate in the future? Thanks!